### PR TITLE
managed nodes with lifecycle nodes

### PR DESCRIPTION
**Pull request by Nicaus**
to view lifecycle node state, execute: 
```
ros2 lifecycle get /[node_name]
```
if it says active, everything should be running as usual

To kill nodes, execute: 

```
ros2 lifecycle set /[node_name] shutdown
```

### CODE REVIEW

File: src/arm_ik/arm_ik/IKNode.py [hunk 1]
```diff
@@ -377,19 +391,27 @@ def validAngles(self, angles):
...
-    try:
-        ik_node.publish_joint_state()
-        loop_rate.sleep()
-    except KeyboardInterrupt:
-        print("Node shutting down due to shutting down node.")
-        break
+  # Spin in a separate thread
+  # thread = threading.Thread(target=rclpy.spin, args=(ik_node, ), daemon=True)
+  # thread.start()
+
```
Comment (vexandmore) [inline src/arm_ik/arm_ik/IKNode.py:404]: remove commented out code

File: src/beep_autonomy/config/test-ekf.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    frequency: 30.0
+    two_d_mode: true # Recommended to use 2d mode for nav2 in mostly planar environments
+    print_diagnostics: true
+    debug: false
+    publish_tf: true
+
+    map_frame: map
+    odom_frame: odom
+    base_link_frame: base_footprint # the frame id used by the turtlebot's diff drive plugin
+    world_frame: odom
```
Comment (vexandmore) [inline src/beep_autonomy/config/test-ekf.yaml:13]: I must've added this and test.launch.py by accident; I'll delete them

File: src/gps/src/gpsPublisherNode.h [hunk 3]
```diff
@@ -16,18 +28,25 @@ struct gpsData
...
 
-class gpsPublisherNode : public rclcpp::Node
-{
- public:
-	explicit gpsPublisherNode(const std::string& gpsTopic = "/gps_data");
+class gpsPublisherNode : public rclcpp_lifecycle::LifecycleNode {
+	public:
+		explicit gpsPublisherNode();
+		callbackReturn on_configure(const rclcpp_lifecycle::State &);
+        callbackReturn on_activate(const rclcpp_lifecycle::State & state);
```
Comment (vexandmore) [inline src/gps/src/gpsPublisherNode.h:35]: inconsistent indenting

File: src/gps/src/gpsServiceClientNode.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            return;
+        }
+        if (!gps_client->get_state()) {
+            return;
+        }
+    }
+
+    // and finally shutdown
+    {
+        time_between_state_changes.sleep();
```
Comment (vexandmore) [inline src/gps/src/gpsServiceClientNode.cpp:214]: What's hapenning here? Why is is calling the shutdown transition?

File: src/service_client/src/service_client.cpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            return;
+        }
+        if (!client->get_state()) {
+            return;
+        }
+    }
+
+    // // and finally shutdown
+    // {
+    //     time_between_state_changes.sleep();
```
Comment (vexandmore) [inline src/service_client/src/service_client.cpp:166]: commented out code

File: src/sil/src/SIL_color_publisher_node.cpp [hunk 6]
```diff
@@ -2,65 +2,142 @@
...
 
-class Configuration : public rclcpp::Node
-{
-public:
-  Configuration()
-  : Node("sil_node")
-  {
-    
-    this->declare_parameter("sil_path", "/dev/ttyUSB1");
+    void callback(){
```
Comment (vexandmore) [inline src/sil/src/SIL_color_publisher_node.cpp:26]: what is the point of this callback?

File: src/wheels_controller/src/wheels_controller_node.cpp [hunk 7]
```diff
@@ -1,6 +1,27 @@
 #include "wheels_controller_node.h"
 
-WheelsControllerNode::WheelsControllerNode(): Node("wheels_controller") {
+WheelsControllerNode::WheelsControllerNode() : LifecycleNode("wheels_controller_node"){}
+// explicit WheelsControllerNode(const std::string& node_name, bool intra_process_comms = false)
+// : rclcpp_lifecycle::LifecycleNode(node_name, 
+// rclcpp::NodeOptions().use_intra_process_comms(intra_process_comms)){};
+
+// void topic_callback(){
```
Comment (vexandmore) [inline src/wheels_controller/src/wheels_controller_node.cpp:8]: remove commented out code

**Review by vexandmore — CHANGES_REQUESTED** [review]

File: src/sil/src/SIL_color_publisher_node.cpp (hunk 6 above)
Comment (Nicaus) [reply src/sil/src/SIL_color_publisher_node.cpp:26]: for the timer

File: src/gps/src/gpsServiceClientNode.cpp (hunk 4 above)
Comment (Nicaus) [reply src/gps/src/gpsServiceClientNode.cpp:214]: This was what the service client was going to be like, (individual files for each node),
I just forgot to delete this file.

File: launch/robot_wheels.py [hunk 8]
```diff
@@ -1,14 +1,49 @@
...
+        name='wheels_controller_node',
+        output='screen',
+        parameters=[
+            {'multiplier': 2000},
+            {'local_mode': True}
+        ],
+        namespace='/',
+    )
+
+    configure_event = RegisterEventHandler(
```
[the remark this answers is not in the document]
Comment (Nicaus) [reply launch/robot_wheels.py:32]: This isn't code I wrote so you'll have to talk to whoever did,
I won't touch it.

File: src/wheels_controller/src/wheels_controller_node.h [hunk 9]
```diff
@@ -93,4 +107,61 @@ private :
    //navigation_goal_status_sub_ = node->create_subscription<action_msgs::msg::GoalStatusArray>;
 };
 
+// class WheelsControllerNode : public rclcpp::Node{
+// public:
+//     WheelsControllerNode();
+//     void TwistMessageCallback(const geometry_msgs::msg::Twist::SharedPtr twist_msg);
+//     void JoyMessageCallback(const sensor_msgs::msg::Joy::SharedPtr joy_msg);
+
+// private :
```
[the remark this answers is not in the document]
Comment (vexandmore) [reply src/wheels_controller/src/wheels_controller_node.h]: Why did you mark as resolved? The commented out code is still here

File: src/sil/src/SIL_color_publisher_node.cpp (hunk 6 above)
Comment (vexandmore) [reply src/sil/src/SIL_color_publisher_node.cpp:26]: Why do you want to call this function once per second? The comments say "for demo purposes", I don't think this is actually needed

File: launch/robot_wheels.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (vexandmore) [reply launch/robot_wheels.py]: I've decided to keep it, but with some unnecessary code removed

File: src/sil/src/SIL_color_publisher_node.cpp (hunk 6 above)
Comment (Nicaus) [reply src/sil/src/SIL_color_publisher_node.cpp:26]: turns out we don't necessarily need a timer

**Review by vexandmore — APPROVED** [review]

File: src/arm_controller/src/arm_controller_node.cpp [hunk 10]
```diff
@@ -177,16 +227,25 @@ void ArmControllerNode::JoyMessageCallback(const sensor_msgs::msg::Joy::SharedPt
...
 {
     rclcpp::init(argc, argv);
-    rclcpp::spin(std::make_shared<ArmControllerNode>());
+    rclcpp::executors::SingleThreadedExecutor exe;
+
+    std::shared_ptr<ArmControllerNode> arm_controller_node = std::make_shared<ArmControllerNode>();
+    exe.add_node(arm_controller_node->get_node_base_interface());
+
+
+    // rclcpp::spin(std::make_shared<ArmControllerNode>());
```
Comment (OZiad) [inline src/arm_controller/src/arm_controller_node.cpp:247]: delete

File: src/gps/src/gpsPublisherNode.h [hunk 11]
```diff
@@ -16,18 +28,25 @@ struct gpsData
...
+		void publishGpsData();
+		gpsData extractGpsData();
+		const float MULTIPLYING_FACTOR = 1e-7;
 
- private:
-	void publishGpsData();
-	gpsData extractGpsData();
-	const float MULTIPLYING_FACTOR = 1e-7;
+		// std::shared_ptr<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::String>> pub_;
+		// std::shared_ptr<rclcpp::TimerBase> timer_;
```
Comment (OZiad) [inline src/gps/src/gpsPublisherNode.h:46]: Delete

File: src/wheels_controller/src/wheels_controller_node.cpp [hunk 12]
```diff
@@ -1,6 +1,8 @@
 #include "wheels_controller_node.h"
 
-WheelsControllerNode::WheelsControllerNode(): Node("wheels_controller") {
+WheelsControllerNode::WheelsControllerNode() : LifecycleNode("wheels_controller_node"){}
+
+callbackReturn WheelsControllerNode::on_configure(const rclcpp_lifecycle::State &){
     this->declare_parameter("can_path", "can0");
     this->declare_parameter("multiplier", 500);
```
Comment (OZiad) [inline src/wheels_controller/src/wheels_controller_node.cpp:7]: Move param declaration to constructor

**Review by OZiad — CHANGES_REQUESTED** [review]

File: src/absenc_interface/src/absenc.h [hunk 13]
```diff
@@ -47,10 +54,16 @@ class AbsencDriver {
     static ABSENC_Error_t ClosePort(int s_fd);
 };
 
-class Absenc: public rclcpp::Node{
+class Absenc: public rclcpp_lifecycle::LifecycleNode{
 public:
     Absenc();
     ~Absenc();
+    callbackReturn on_configure(const rclcpp_lifecycle::State &);
```
[the remark this answers is not in the document]
Comment (Nicaus) [reply src/absenc_interface/src/absenc.h:61]: what do you mean by that?

**Review by OZiad — APPROVED** [review]

**Pull request merged**